crates/types/src/partition_table.rs

```diff
@@ -93,21 +86,120 @@ impl FixedPartitionTable {
         self.version = version;
     }
 
-    pub fn partitioner(&self) -> Partitioner {
-        Partitioner::new(self.num_partitions)
+    pub fn partitions(&self) -> impl Iterator<Item = (&PartitionId, &Partition)> {
+        self.partitions.iter()
     }
 
-    fn partition_key_to_partition_id(
-        num_partitions: u64,
+    pub fn num_partitions(&self) -> u64 {
+        u64::try_from(self.partitions.len()).expect("number of partitions should fit into u64")
+    }
+
+    pub fn get_partition(&self, partition_id: &PartitionId) -> Option<&Partition> {
+        self.partitions.get(partition_id)
+    }
+}
+
+impl Versioned for PartitionTable {
+    fn version(&self) -> Version {
+        self.version
+    }
+}
+
+flexbuffers_storage_encode_decode!(PartitionTable);
+
+impl FindPartition for PartitionTable {
+    fn find_partition_id(
+        &self,
         partition_key: PartitionKey,
-    ) -> PartitionId {
-        let num_partitions = u128::from(num_partitions);
-        let partition_key = u128::from(partition_key);
+    ) -> Result<PartitionId, PartitionTableError> {
+        // partition key ranges are inclusive, so let's look for the next partition key boundary >=
+        // partition_key to find the owning partition
+        self.partition_key_index
+            .range(partition_key..)
+            .next()
+            .map(|(_, partition_id)| *partition_id)
+            .ok_or(PartitionTableError(partition_key))
+    }
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, serde::Serialize, serde::Deserialize)]
+pub struct Partition {
+    pub key_range: RangeInclusive<PartitionKey>,
+    pub leader: Option<GenerationalNodeId>,
+    pub node_set: BTreeSet<GenerationalNodeId>,
```

Comment:
I'm not sure whether those should exist here. I think operational status and deployment configuration for partitions should be store separately (those are cluster controller's metadata)

In the future we will need to have a `state` for the partition so we can archive old partitions which won't appear in the index but can still be addressed by partition_id. But that's something we can add when we need it. If you'd like to take the first step in this, maybe we can have an enum with `Active` / `Archived` states.

Reply:
The idea behind the `leader` field was that maybe some components like the `ingress` would like to know who the current leader is for a given partition to query that leader (e.g. when submitting an invocation or retrieving an invocation result). The `node_set` is not so helpful, agreed. I can remove both fields until we have a concrete need for it.